### DISCUSSION TITLE
Is there a limit to the amount of rows returned?

### DISCUSSION THREAD

**Discussion by Zloka** [supabase/supabase#3765] [Questions] [answered]
The [documentation says the maximum rows returned defaults to 1000](https://supabase.io/docs/reference/javascript/select#:~:text=By%20default%2C%20Supabase%20projects%20will%20return%20a%20maximum%20of%201%2C000%20rows.). We have increased this limit, and for our use-case we may need to increase it further in the future.

I was wondering if there is some maximum value, or if I can just arbitrarily increase this limit?

**Comment by silentworks** [comment, marked as answer]
You should probably keep that limit and paginate the rest of the results. The higher you increase it to the slower the query will become to return the records. I don't think there is a maximum value.

**Reply by Zloka** [reply]
Well, as I said, we have a specific use case where we need all the rows the query returns (basically a data dump). In other cases, for example when users view the data, it is paginated.

So if the only downside is increased query time, for my data dump use-case that doesn't really matter and I would rather save the overhead of adding pagination, hence why I wanted to know about the limit :)

**Reply by soedirgo** [reply]
There isn't a max value for the limit - on the PostgREST side (which is what's under the hood) [the default value is no limit](https://postgrest.org/en/v8.0/configuration.html?highlight=max-rows#configuration), though I can't confirm how you can set ∞ from the dashboard.

**Reply by Zloka** [reply]
Alright, thank you! That sounds good enough for my purposes, I can just set a high enough threshold then, as my data does not grow very quickly and it will be trivial to increase if I hit it again 🙂

**Reply by largesseio** [reply]
The setting is being ignored on the API UI for the increase. I need a higher limit than 1000. Please Help

**Reply by soedirgo** [reply]
Hey @largesseio, this is a regression: https://github.com/supabase/supabase/issues/4544

**Reply by pencilcheck** [reply]
really, the limit is 1000?

**Reply by soedirgo** [reply]
@pencilcheck it's the default limit - you can increase it.

**Reply by ssmcb** [reply]
Is there any way of setting the max rows limit by table instead of the whole API?

**Reply by GaryAustin1** [reply]
No.  You only have the option like .limit() in the calls on a table.

**Comment by wscourge** [comment]
The accepted answer doesn't help at all. You can increase the limit here:

https://supabase.com/dashboard/project/:id/settings/api

at the very bottom (Search "Max rows"). Maximum value allowed is `1000000`.

![image](https://github.com/supabase/supabase/assets/16985871/16667ed8-e355-45f5-a9e5-f2871973bb57)

**Reply by iwasrobbed** [reply]
Doesn't seem like there's a localhost variant of this for local dev. Any chance there is and I just can't find it?

**Reply by wscourge** [reply]
Add `[api].max_rows = 1000000` to your _supabase/config.toml_:

```
# A string used to distinguish different Supabase projects on the same host. Defaults to the working
# directory name when running `supabase init`.
project_id = "demo"

[api]
# Port to use for the API URL.
port = 54321
# Schemas to expose in your API. Tables, views and stored procedures in this schema will get API
# endpoints. public and storage are always included.
schemas = ["public", "storage", "graphql_public"]
# Extra schemas to add to the search_path of every request. public is always included.
extra_search_path = ["public", "extensions"]
# The maximum number of rows returns from a view, table, or stored procedure. Limits payload size
# for accidental or malicious requests.
max_rows = 1000000

[db]
# Port to use for the local database URL.
port = 54322

...
...
..
```

**Reply by celine-m-s** [reply]
Thank you so much for this answer. I've been struggling for hours with the Supabase's JS client before realizing my problem was probably due to this limitation.

**Reply by pauluk** [reply]
I found this setting in Data API max rows there 
<img width="314" height="518" alt="image" src="https://github.com/user-attachments/assets/d7747b2b-5089-454e-a759-961e3531636d" />

**Reply by rioalba** [reply]
> Add `[api].max_rows = 1000000` to your _supabase/config.toml_:
> 
> ```
> # A string used to distinguish different Supabase projects on the same host. Defaults to the working
> # directory name when running `supabase init`.
> project_id = "demo"
> 
> [api]
> # Port to use for the API URL.
> port = 54321
> # Schemas to expose in your API. Tables, views and stored procedures in this schema will get API
> # endpoints. public and storage are always included.
> schemas = ["public", "storage", "graphql_public"]
> # Extra schemas to add to the search_path of every request. public is always included.
> extra_search_path = ["public", "extensions"]
> # The maximum number of rows returns from a view, table, or stored procedure. Limits payload size
> # for accidental or malicious requests.
> max_rows = 1000000
> 
> [db]
> # Port to use for the local database URL.
> port = 54322
> 
> ...
> ...
> ..
> ```

This is the solution for local development with Supabase Start. For projects hosted on Supabase, you need to change this in the Dashboard:

  Supabase Dashboard → Project Settings → API → Max Rows

Use pagination, which completely bypasses this limit! Instead of requesting 100,000 rows in a query, do the following:

Batch 1: fetch rows 0-999 → 1,000 results
Batch 2: fetch rows 1,000-1,999 → 1,000 results
...
